feat: Add quay plugin blog (#21)
janus-idp/janus-idp.github.io: apps/landing-page/pages/blog/exploring-quay-registry-in-backstage.mdx

---
title: Exploring Quay registry in Backstage
date: 2023/02/20

> description: The Janus IDP family of Backstage plugins is expanding! Please welcome our new member - a frontend plugin that enriches application view in Backstage with insights from Quay registry.

Reviewer: ```suggestion
description: The Janus IDP family of Backstage plugins is expanding! Please welcome our new member - a frontend plugin that enriches application view in Backstage with insights from a Quay hosted registry.
```

---

import { Authors } from 'ui/components';

# Exploring Quay registry in Backstage

<Authors authors={['tumido_']} />

> The Janus IDP family of Backstage plugins is expanding! Please welcome our new member - a frontend plugin that enriches application view in Backstage with insights from the Quay registry.

Reviewer: ```suggestion
The Janus IDP family of Backstage plugins is expanding! Please welcome our new member - a frontend plugin that enriches application view in Backstage with insights from a Quay hosted registry.
```

> Backstage charts and models the software ecosystem as Catalog entities. Users compose small individual service components into a bigger picture. And in order to truly understand and fully describe the individual building blocks, Backstage users construct views to capture different aspects of these components. From technical documentation, through dependencies and relation to deployment state, CI and CD. And part of this picture is understanding what is actually being deployed into live environments: a view of the available and deployed container images. This new Quay plugin for Backstage brings that view.

Reviewer: Is "charts" somewhat overloading? maybe just drop that adjective as it could be confused with Helm?

Author: You may be right, I agree.

> Backstage charts and models the software ecosystem as Catalog entities. Users compose small individual service components into a bigger picture. And in order to truly understand and fully describe the individual building blocks, Backstage users construct views to capture different aspects of these components. From technical documentation, through dependencies and relation to deployment state, CI and CD. And part of this picture is understanding what is actually being deployed into live environments: a view of the available and deployed container images. This new Quay plugin for Backstage brings that view.

Reviewer: ```suggestion
Backstage charts and models the software ecosystem as Catalog entities. Users compose small individual service components into a bigger picture. And in order to truly understand and fully describe the individual building blocks, Backstage users construct views to capture different aspects of these components; from technical documentation, through dependencies and relation to deployment state, CI and CD. And part of this picture is understanding what is actually being deployed into live environments. In many cases, the deployment artifact is a container image and the desire to view all of the available and deployed container images. This new Quay plugin for Backstage brings in that capability.
```

Quay is an OCI-compatible registry that allows users to build, store and distribute container images and other OCI artifacts. It is available as a hosted service on quay.io as well as a self-hosted environment deployable to any OpenShift cluster through a Quay operator. 

## Installation and setup

> With this plugin, viewing available container images in a particular repository is easy. The following guide will elaborate in detail on individual steps and all currently supported settings and options.

Reviewer: ```suggestion
With this plugin, viewing available container images in a particular repository is easy. The following guide will elaborate in detail on individual steps for enabling the integration and all currently supported settings and options.
```

First, it is necessary to install the plugin. Please add it to the frontend of your Backstage instance:

```bash
yarn workspace app add @janus-idp/backstage-plugin-quay
```

### Connecting to Quay registry via Proxy

> Plugin leverages the Backstage native proxy capabilities to query Quay API; therefore, it needs to be configured in `app-config.yaml`. In order to connect the Backstage instance to the public hosted Quay.io environment, the following configuration can be used:

Reviewer: ```suggestion
The plugin leverages the Backstage native proxy capabilities to query the Quay API; therefore, configurations need to be made to the `app-config.yaml`. In order to connect the Backstage instance to the public hosted Quay.io environment, the following configuration can be used:
```

```yaml filename="app-config.yaml"
proxy:
  '/quay/api':
    target: 'https://quay.io'
    changeOrigin: true
    headers:
      X-Requested-With: 'XMLHttpRequest'
```

> When accessing private Quay repositories, it is necessary to extend this configuration with `Authorization` header and Quay API token. This token can be obtained by creating a Quay Application using the [steps outlined in this documentation](https://docs.quay.io/api/#applications-and-tokens). Once a token is obtained, the header can be set by extending the `app-config.yaml` setup above with an additional header:

Reviewer: ```suggestion
When accessing private Quay repositories, it may be necessary to extend this configuration with an `Authorization` header and Quay API token. This token can be obtained by creating a Quay Application using the [steps outlined in this documentation](https://docs.quay.io/api/#applications-and-tokens). Once a token is obtained, the header can be set by extending the `app-config.yaml` setup above with an additional header:
```

```yaml {7} filename="app-config.yaml"
proxy:
  '/quay/api':
    target: 'https://quay.io'
    changeOrigin: true
    headers:
      X-Requested-With: 'XMLHttpRequest'
      Authorization: 'Bearer ${QUAY_TOKEN}'
```

Be aware that the `QUAY_TOKEN` is an environment variable that has to be available to the Backstage instance at runtime.

> Another popular demand may be to use this plugin to access a private registry, a self-hosted Quay deployment. This can be achieved by simply changing the `target` property in the settings above. In addition to that, if the self-hosted Quay registry is also deployed with a self-signed certificate that is not in the certificate chain of trust for the Backstage instance, the `secure` option has to be unset.

Reviewer: ```suggestion
Another popular option is to be able to target a self-hosted Quay deployment. This can be achieved by simply changing the `target` property in the settings above with the location of the Quay instance. In addition, if the self-hosted Quay registry is also deployed with a certificate that is not in the certificate chain of trust for the Backstage instance, the `secure` option has to be unset.
```

Reviewer: Would be a good feature of the plugin to support specifying the location of a certificate that can be used to support a non-public CA

Author: It's using https://www.npmjs.com/package/http-proxy-middleware behind the scenes. It seems like it doesn't allow you to specify custom CA in any way. But you may be able to mount it to the container maybe? Anyway we may want to explore that in future.

```yaml {3,8} filename="app-config.yaml"
proxy:
  '/quay/api':
    target: '<SELF_HOSTED_QUAY_ENDPOINT>'
    changeOrigin: true
    headers:
      X-Requested-With: 'XMLHttpRequest'
      Authorization: 'Bearer ${QUAY_TOKEN}'
    secure: [true|false]
```

More details on available Backstage proxy settings can be found in the [upstream documentation](https://backstage.io/docs/plugins/proxying). 

> This plugin conforms to the pattern used by many other Backstage plugins that use Backstage proxy and provides a mechanism to change the default proxy path `/quay/api` via the following `app-config.yaml` settings, if needed:

Reviewer: ```suggestion
This plugin conforms to the pattern used by many other Backstage plugins that use the Backstage proxy and provides a mechanism to change the default proxy path `/quay/api` via the following `app-config.yaml` settings, if needed:
```

```yaml filename="app-config.yaml"
quay:
  proxyPath: /custom/quay/path
```

### Enabling Quay plugin widget in UI

Now that the plugin is configured to access the desired Quay registry, enable the UI by enabling an additional view in the frontend application.(`packages/app/src/components/catalog/EntityPage.tsx` file in the bootstrap application)

```tsx filename="packages/app/src/components/catalog/EntityPage.tsx"
import { QuayPage, isQuayAvailable } from '@fmenesesg/backstage-plugin-quay';

const serviceEntityPage = (
  <EntityPageLayout>
    // ...
    <EntityLayout.Route if={isQuayAvailable} path="/quay" title="Quay">
      <QuayPage />
    </EntityLayout.Route>
  </EntityPageLayout>
);
```

## Using the plugin

> Finally, after the plugin is fully set up, it needs to be advised on what data to display for individual catalog entities. Extend the entity with an annotation, a similar experience Backstage users are used to with other plugins:

Reviewer: ```suggestion
Finally, after the plugin is fully set up, it needs to be instructed on what data to display for individual catalog entities. Extend the entity with an annotation, a similar experience Backstage users are used to with other plugins:
```

```yaml
metadata:
  annotations:
    'quay.io/repository-slug': '<ORGANIZATION>/<REPOSITORY>'
```

> For example if we annotate a `Component` with `'quay.io/repository-slug': 'janus-idp/redhat-backstage-build'`, we are presented with following page:

Reviewer 2: ```suggestion
For example, if we annotate a `Component` with `'quay.io/repository-slug': 'janus-idp/redhat-backstage-build'`, we are presented with the following page:
```

![Backstage view for janus-idp/redhat-backstage-build Quay repository](/images/blog/exploring-quay-registry-in-backstage/widget.png)

## Next steps

Although this plugin doesn’t have the vast features available in the Quay UI, it brings much-needed value to Backstage users.  In the future, we plan to iterate on this plugin and provide users with more insights into unique Quay functions like vulnerability scanning and detailed manifest views.

> To contribute to this plugin, report issues, seek guidance or provide feedback, please see our github repository [https://github.com/janus-idp/backstage-plugins/tree/main/plugins/quay](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/quay) and/or reach out to us on Janus IDP Slack.

Reviewer: ```suggestion
To contribute to this plugin, report issues, seek guidance or provide feedback, please see our GitHub repository [https://github.com/janus-idp/backstage-plugins/tree/main/plugins/quay](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/quay) and/or reach out to us on [Janus IDP Slack](https://join.slack.com/t/janus-idp/shared_invite/zt-1nii16o6e-SGscZ4YtAktL6rRtZZBUfA).
```

Quay is a OCI-compatible registry that allows users to build, store and distribute container images and other OCI artifacts. It is available as a hosted service on quay.io as well as a self hosted environment deployable to any OpenShift cluster through a Quay operator. 

## Installation and setup

With this plugin, viewing available container images in a particular repository is easy with this plugin. The following guide will elaborate in detail individual steps and all currently supported settings and options.

First, it is necessary to install the plugin. Please add it to the frontend of your Backstage instance:

```
yarn workspace app add @janus-idp/backstage-plugin-quay
```

### Connecting to Quay registry via Proxy

Plugin leverages the Backstage native proxy capabilities to query Quay API, therefore it needs to be configured in `app-config.yaml`. In order to connect the Backstage instance to the public hosted Quay.io environment, following configuration can be used:

> ```yaml

Reviewer 3: ```suggestion
```yaml filename="app-config.yaml"
```

proxy:
  '/quay/api':
    target: 'https://quay.io'
    changeOrigin: true
    headers:
      X-Requested-With: 'XMLHttpRequest'
```

When accessing private Quay repositories, it is necessary to extend this configuration with `Authorization` header and Quay API token. This token can be obtained by creating a Quay Application using [steps outlined in this documentation](https://docs.quay.io/api/#applications-and-tokens). Once a token is obtained, the header can be set by extending the `app-config.yaml` setup above with additional header:

> ```diff

Author: Is there a way how can I highlight an addition in YAML?

The purpose of this snippet is to show that you need to add 1 line at this exact place.

> ```diff

Author: Thank you.. that did the trick. FTR, I've managed to educate myself more here: https://nextra.site/docs/guide/syntax-highlighting

```

Be aware that the `QUAY_TOKEN` is an environment variable that has to be available to the Backstage instance at runtime.

Another popular demand may be to use this plugin to access a private registry, a self-hosted Quay deployment. This can be achieved by a simple change of the `target` property in the settings above. In addition to that, if the self hosted Quay registry is also deployed with a self-signed certificate that is not in the certificate chain of trust for the Backstage instance, the `secure` option has to be unset.

```diff
  proxy:
    '/quay/api':
-     target: 'https://quay.io'
+     target: '<SELF_HOSTED_QUAY_ENDPOINT>'
      changeOrigin: true
      headers:
        X-Requested-With: 'XMLHttpRequest'
        Authorization: 'Bearer ${QUAY_TOKEN}'
+     secure: true/false
```

More details on available Backstage proxy settings can be found in the [upstream documentation](https://backstage.io/docs/plugins/proxying). 

This plugin conforms to the pattern used by many other Backstage plugins that use Backstage proxy and provides a mechanism to change the default proxy path `/quay/api` via the following `app-config.yaml` settings, if needed:

> ```yaml

Reviewer 3: ```suggestion
```yaml filename="app-config.yaml"
```

quay:
  proxyPath: /custom/quay/path
```

### Enabling Quay plugin widget in UI

Now that  the plugin is configured to access the desired Quay registry, enable the UI by enabling an additional view in the frontend application.(`packages/app/src/components/catalog/EntityPage.tsx` file in the bootstrap application)

> ```tsx

Reviewer 3: ```suggestion
```tsx filename="packages/app/src/components/catalog/EntityPage.tsx"
```
